src/registration/metric/demons.h

```diff
@@ -47,6 +51,7 @@ namespace MR
           }
 
           ~Demons () {
+            std::lock_guard<std::mutex> lock (*mutex);
```

Comment:
I've got a feeling that this kind of issue first popped up [here](https://github.com/MRtrix3/mrtrix3/blob/master/src/dwi/tractography/seeding/base.h#L61-L86) (fixed [here](https://github.com/MRtrix3/mrtrix3/commit/1804651b47cf68026c6dbba2c2b07b022d96f23d#diff-b9c71cd67db8628819b6656e0b04b5ea)): the volumes of seed ROIs were being under-estimated, which was resulting in incomplete coverage of seed points when using e.g. `-seed_random_per_voxel`. That example doesn't involve class destructors, but I *think* there was at least one subsequent example where a similar effect was observed in the case of class destructors (it's just less likely to arise because potential race conditions only occur on destruction rather than every single increment). Having trouble remembering / finding an example though - suspect it may have been from SIFT2 development, which doesn't have a full Git history.